cirq-core/cirq/transformers/merge_single_qubit_gates.py

```diff
@@ -152,3 +156,257 @@ def merge_func(m1: 'cirq.Moment', m2: 'cirq.Moment') -> Optional['cirq.Moment']:
+    atol: float,
+) -> Tuple[List['cirq.Circuit'], frozenset[str], frozenset[str]]:
+    """Helper function to merge single qubit ops of resolved circuits to PhasedXZ ops
+      using merge_k_qubit_unitaries.
+
+    Args:
+        resolved_circuits: A list of circuits where symbols have been replaced with concrete values.
+        symbolized_single_tag: The tag applied to single-qubit operations that originally
+          contained symbols before parameterizations.
+
+    Returns:
+        Tuple of merge_counts, merged_circuits, and merge_tags, where
+        merged ops in merged_circuits are tagged by merge_tags.
+    """
+    merge_counts: list[int] = []  # number of merges per resolved_circuit
+    merged_circuits: list['cirq.Circuit'] = []
+    tag_iter: itertools.count
+    phxz_tag_prefix = "_phxz"
+
+    def rewriter(circuit_op: 'cirq.CircuitOperation') -> 'cirq.OP_TREE':
+        nonlocal tag_iter
+        tag: Optional[str] = None
+
+        u = protocols.unitary(circuit_op)
+        if protocols.num_qubits(circuit_op) == 0:
+            return ops.GlobalPhaseGate(u[0, 0]).on()
+        # If any of the op in the merged circuit_op is a symbolized single qubit gate,
+        # tag the merged phxz gate with next tag id, for further parameterization references.
+        for op in circuit_op.circuit.all_operations():
+            if symbolized_single_tag in op.tags:
+                tag = f"{phxz_tag_prefix}_{next(tag_iter)}"
+                break
+        gate = single_qubit_decompositions.single_qubit_matrix_to_phxz(u, atol) or ops.I
+        op = gate.on(circuit_op.qubits[0])
+        return op.with_tags(tag) if tag else op
+
+    for resolved_circuit in resolved_circuits:
+        tag_iter = itertools.count(start=0, step=1)
+        merged_circuits.append(
+            merge_k_qubit_gates.merge_k_qubit_unitaries(
+                resolved_circuit, k=1, context=context, rewriter=rewriter
+            )
+        )
+        merge_counts.append(next(tag_iter))
+
+    if not all(count == merge_counts[0] for count in merge_counts):
+        raise RuntimeError("Different resolvers in sweep resulted in different merged structures.")
+
+    merge_tags: frozenset[str] = frozenset(
+        {f"{phxz_tag_prefix}_{i}" for i in range(merge_counts[0])}
+    )
+    new_symbols: frozenset[str] = frozenset(
+        set().union(*[{f"x{i}", f"z{i}", f"a{i}"} for i in range(merge_counts[0])])
+    )
+
+    return merged_circuits, merge_tags, new_symbols
+
+
+def _get_merge_tag_id(merge_tags: frozenset[str], op_tags: Tuple[Hashable, ...]) -> Optional[str]:
+    """Extract the id `i` from the merge tag `_phxz_i` if it exists."""
+    the_merge_tag: set[str] = set(merge_tags.intersection(op_tags))
+    if len(the_merge_tag) == 0:
+        return None
+    if len(the_merge_tag) > 1:
+        raise RuntimeError("Multiple merge tags found.")
+    return the_merge_tag.pop().split("_")[-1]
+
+
+def _map_merged_ops_to_symbolized_phxz(
```

Comment:
this should be a public transformer method

Reply:
Done in symbolize.py .